src/Translation_unit.ml

```diff
@@ -114,94 +108,95 @@ let parse parse_ast (conf : Conf.t) ic =
     Location.warning_printer := warning_printer ;
     raise e
 
-(** Debug: dump internal ast representation to file. *)
-let dump xunit dir base suf ext ast =
-  if Conf.debug then (
-    let tmp = Filename.concat dir (base ^ suf ^ ext) in
-    let oc = Out_channel.create tmp in
-    xunit.printast (Caml.Format.formatter_of_out_channel oc) ast ;
-    Out_channel.close oc )
-
-type result =
-  | Ok
-  | Invalid_source of exn
-  | Unstable of int
-  | Ocamlformat_bug of exn
+type error =
+  | Invalid_source of {input_name: string; exn: exn}
+  | Unstable of
+      { input_name: string
+      ; iteration: int
+      ; prev: string
+      ; next: string }
+  | Ocamlformat_bug of {input_name: string; exn: exn}
   | User_error of string
 
 let ellipsis n msg =
   let msg = String.strip msg in
   if n > 0 && String.length msg > (n * 2) + 10 then
-    Format.sprintf "%s ... %s" (String.prefix msg n) (String.suffix msg n)
+    sprintf "%s ... %s" (String.prefix msg n) (String.suffix msg n)
   else msg
 
 let ellipsis_cmt = ellipsis 50
 
-let parse_print (XUnit xunit) (conf : Conf.t) ~input_name ~input_file ic
-    ofile =
+let make_dump xunit input_name dir =
   let dir =
-    match ofile with
-    | Some ofile -> Filename.dirname ofile
-    | None -> Filename.get_temp_dir_name ()
+    match dir with Some dir -> dir | None -> Filename.get_temp_dir_name ()
   in
-  let base = Filename.(remove_extension (basename input_name)) in
   let ext = Filename.extension input_name in
+  let base = Filename.remove_extension (Filename.basename input_name) in
+  let with_file suf ext f =
+    let tmp = Filename.concat dir (base ^ suf ^ ext) in
+    Out_channel.with_file tmp ~f ;
+    tmp
+  in
+  let dump_ast suf ast =
+    if Conf.debug then
+      let ext = ".ast" in
+      let (_filename : string) =
+        with_file suf ext (fun oc ->
+            xunit.printast (Caml.Format.formatter_of_out_channel oc) ast )
+      in
+      ()
+  in
+  let dump_formatted suf fmted =
+    if Conf.debug then
+      let file =
+        with_file suf ext (fun oc -> Out_channel.output_string oc fmted)
+      in
+      [("output file", String.sexp_of_t file)]
+    else []
+  in
+  (dump_ast, dump_formatted)
+
+let format xunit (conf : Conf.t) ~input_name ~source ~parsed ~output_file =
+  Location.input_name := input_name ;
+  let dump_ast, dump_formatted =
```

Comment:
Can we go further and make `dump_ast` and `dump_formatted` parameters of the `format` function?

Reply:
I just did that. The API doesn't look great. I don't really know how to make it better.

Reply:
@gpetiot, how do you feel about taking over that PR ? I don't have much time to spend on this. Feel free to push to the branch directly.

Reply:
@hhugo can you allow contributors to push to this branch? Right now I can't, thanks!